src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -182,31 +183,63 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
+        private static Socket Connect(string serverName, int port, TimeSpan timeout)
         {
-            IPAddress[] addresses = await Dns.GetHostAddressesAsync(serverName).ConfigureAwait(false);
-            IPAddress targetAddrV4 = Array.Find(addresses, addr => (addr.AddressFamily == AddressFamily.InterNetwork));
-            IPAddress targetAddrV6 = Array.Find(addresses, addr => (addr.AddressFamily == AddressFamily.InterNetworkV6));
-            if (targetAddrV4 != null && targetAddrV6 != null)
+            IPAddress[] ipAddresses = Dns.GetHostAddresses(serverName);
+            IPAddress serverIPv4 = null;
+            IPAddress serverIPv6 = null;
+            foreach (IPAddress ipAdress in ipAddresses)
             {
-                return await ParallelConnectAsync(new IPAddress[] { targetAddrV4, targetAddrV6 }, port).ConfigureAwait(false);
+                if (ipAdress.AddressFamily == AddressFamily.InterNetwork)
+                {
+                    serverIPv4 = ipAdress;
+                }
+                else if (ipAdress.AddressFamily == AddressFamily.InterNetworkV6)
+                {
+                    serverIPv6 = ipAdress;
+                }
             }
-            else
-            {
-                IPAddress targetAddr = (targetAddrV4 != null) ? targetAddrV4 : targetAddrV6;
-                var socket = new Socket(targetAddr.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+            ipAddresses = new IPAddress[] { serverIPv4, serverIPv6 };
+            Socket[] sockets = new Socket[2];
 
-                try
+            CancellationTokenSource cts = new CancellationTokenSource();
+            cts.CancelAfter(timeout);
+            void Cancel()
+            {
+                foreach (Socket socket in sockets)
                 {
-                    await socket.ConnectAsync(targetAddr, port).ConfigureAwait(false);
+                    if (socket != null && !socket.Connected)
+                    {
+                        socket.Dispose();
+                    }
                 }
-                catch
+            }
+            cts.Token.Register(Cancel);
+
+            Socket availableSocket = null;
+            for (int i = 0; i < sockets.Length; ++i)
+            {
+                try
                 {
-                    socket.Dispose();
-                    throw;
+                    if (ipAddresses[i] != null)
+                    {
+                        sockets[i] = new Socket(ipAddresses[i].AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+                        sockets[i].Connect(ipAddresses[i], port);
+                        if (sockets[i].Connected)
+                        {
+                            availableSocket = sockets[i];
+                            break;
+                        }
+                        else
+                        {
+                            sockets[i].Dispose();
```

Comment:
Could this possibly cause an ObjectDisposedException from calling dispose again in the cancellation function? Maybe null out this socket after dispose?

Reply:
Good point! I will look into it.

Reply:
`socket.Dispose()` can be multiple times, and it does not throw exception. It is still a good idea to null out the socket after disposal since it will make not to run `dispose()'  again.